utils/safeList.js

```diff
@@ -0,0 +1,286 @@
+const xml2js = require('xml2js');
+
+const recoverableErrors = new Set([
+    'TimestampParserError',
+]);
+
+const NO_VALUE = Symbol('noValue');
+
+/**
+ * @typedef {Object} ExtractResult
+ * @property {any} value - extracted value
+ * @property {Error|undefined} error - error if any
+ */
+
+/**
+ * @typedef {function} Extractor
+ * @param {string} key - key to extract
+ * @param {object} data - data to extract from
+ * @returns {ExtractResult} - extracted value
+ */
+
+/**
+ * extractKeys - extract keys from data using the provided extractors
+ *
+ * @param {object} data - data to extract keys from
+ * @param {object<string, Extractor>} keys - Mapping of keys to extractors
+ * @returns {ExtractResult} - extracted values
+ */
+function extractKeys(data, keys) {
+    const extracted = {};
+    for (const [key, extractor] of Object.entries(keys)) {
+        const { value, error } = extractor(key, data);
+        if (error) {
+            return { error };
+        }
+        if (value !== NO_VALUE) {
+            extracted[key] = value;
+        }
+    }
+    return { value: extracted };
+}
+
+/**
+ * safeExtract - extract a value from data
+ * @param {string} key - key to extract
+ * @param {any} data - data to extract from
+ * @returns {ExtractResult} - extracted value
+ */
+function safeExtract(key, data) {
+    if (data[key] && data[key].length > 0) {
```

Comment:
As this is documented as `any`, maybe we should ensure that's an array first?

Reply:
The xml parser only returns an object of arrays but I can swap it for a `Array.isArray` call instead